File: packages/sdk-go/client_test.go
package setu_test

import (
	"context"
	"net/http"
	"net/http/httptest"
	"testing"

	setu "github.com/setu-gateway/setu-gateway-go"
)

func TestChatCompletionsCreate_SendsExpectedRequestAndParsesResponse(t *testing.T) {
	var gotPath, gotAuth string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		gotAuth = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"id":"chatcmpl-1","choices":[{"message":{"role":"assistant","content":"hi back"}}]}`))
	}))
	defer server.Close()

	client := setu.NewClient(setu.WithBaseURL(server.URL), setu.WithAPIKey("sk_setu_test"))
	resp, err := client.Chat.Completions.Create(context.Background(), setu.ChatCompletionRequest{
		Model:    "gpt-4o",
		Messages: []setu.Message{{Role: "user", Content: "hi"}},
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if gotPath != "/v1/chat/completions" {
		t.Errorf("expected path /v1/chat/completions, got %s", gotPath)
	}
	if gotAuth != "Bearer sk_setu_test" {
		t.Errorf("expected Authorization header to be set, got %q", gotAuth)
	}
	if resp["id"] != "chatcmpl-1" {
		t.Errorf("expected response to be parsed, got %v", resp)
	}
}

func TestRequestFailure_ReturnsAPIError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`{"detail":"invalid key"}`))
	}))
	defer server.Close()

	client := setu.NewClient(setu.WithBaseURL(server.URL))
	_, err := client.Models.List(context.Background())
	if err == nil {
		t.Fatal("expected an error, got nil")
	}
	apiErr, ok := err.(*setu.APIError)
	if !ok {
		t.Fatalf("expected *setu.APIError, got %T: %v", err, err)
	}
	if apiErr.StatusCode != http.StatusUnauthorized {
		t.Errorf("expected status 401, got %d", apiErr.StatusCode)
	}
}

func TestConnectionFailure_ReturnsConnectionError(t *testing.T) {
	client := setu.NewClient(setu.WithBaseURL("http://127.0.0.1:1"))
	_, err := client.Models.List(context.Background())
	if err == nil {
		t.Fatal("expected an error, got nil")
	}
	if _, ok := err.(*setu.ConnectionError); !ok {
		t.Fatalf("expected *setu.ConnectionError, got %T: %v", err, err)
	}
}

// TestLiveGatewaySmoke only runs when a real gateway is reachable at the default
// base URL - skipped otherwise rather than failing CI in an environment with no
// gateway running.
func TestLiveGatewaySmoke(t *testing.T) {
	client := setu.NewClient()
	resp, err := client.Chat.Completions.Create(context.Background(), setu.ChatCompletionRequest{
		Model:    "gpt-4o",
		Messages: []setu.Message{{Role: "user", Content: "Say 'sdk-go works' and nothing else."}},
	})
	if err != nil {
		t.Skipf("no live gateway reachable at default base URL, skipping: %v", err)
	}
	if resp["id"] == nil {
		t.Errorf("expected a response with an id field, got %v", resp)
	}
}

File: packages/sdk-rust/tests/client_test.rs
use setu_gateway::{ChatCompletionRequest, Client, Error, Message};
use wiremock::matchers::{header, method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};

#[tokio::test]
async fn chat_completions_create_sends_expected_request_and_parses_response() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/v1/chat/completions"))
        .and(header("Authorization", "Bearer sk_setu_test"))
        .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
            "id": "chatcmpl-1",
            "choices": [{"message": {"role": "assistant", "content": "hi back"}}]
        })))
        .mount(&server)
        .await;

    let client = Client::new().with_base_url(server.uri()).with_api_key("sk_setu_test");
    let resp = client
        .chat()
        .completions()
        .create(ChatCompletionRequest {
            model: "gpt-4o".into(),
            messages: vec![Message { role: "user".into(), content: "hi".into() }],
            temperature: None,
            top_p: None,
            max_tokens: None,
        })
        .await
        .expect("request should succeed");

    assert_eq!(resp["id"], "chatcmpl-1");
}

#[tokio::test]
async fn request_failure_returns_api_error() {
    let server = MockServer::start().await;
    Mock::given(method("GET"))
        .and(path("/v1/models"))
        .respond_with(ResponseTemplate::new(401).set_body_json(serde_json::json!({"detail": "invalid key"})))
        .mount(&server)
        .await;

    let client = Client::new().with_base_url(server.uri());
    let err = client.models().list().await.expect_err("expected an error");
    match err {
        Error::Api { status, body } => {
            assert_eq!(status, 401);
            assert_eq!(body, "invalid key");
        }
        other => panic!("expected Error::Api, got {other:?}"),
    }
}

#[tokio::test]
async fn connection_failure_returns_connection_error() {
    let client = Client::new().with_base_url("http://127.0.0.1:1");
    let err = client.models().list().await.expect_err("expected an error");
    assert!(matches!(err, Error::Connection(_)), "expected Error::Connection, got {err:?}");
}

/// Only runs meaningfully when a real gateway is reachable at the default base URL;
/// otherwise the connection error path itself is still exercised (so the test isn't
/// flaky either way, it just can't prove the live round-trip in that environment).
#[tokio::test]
async fn live_gateway_smoke() {
    let client = Client::new();
    match client
        .chat()
        .completions()
        .create(ChatCompletionRequest {
            model: "gpt-4o".into(),
            messages: vec![Message { role: "user".into(), content: "Say 'sdk-rust works' and nothing else.".into() }],
            temperature: None,
            top_p: None,
            max_tokens: None,
        })
        .await
    {
        Ok(resp) => assert!(resp.get("id").is_some(), "expected a response with an id field, got {resp:?}"),
        Err(Error::Connection(_)) => eprintln!("no live gateway reachable at default base URL, skipping assertion"),
        Err(e) => panic!("unexpected error: {e:?}"),
    }
}
